stan/math/rev/core/operator_addition.hpp

```diff
@@ -116,10 +138,153 @@ inline var operator+(const var& a, Arith b) {
+ * @param b Second variable operand.
+ * @return Variable result of adding two variables.
+ */
+template <typename VarMat1, typename VarMat2,
+          require_all_rev_matrix_t<VarMat1, VarMat2>* = nullptr>
+inline auto operator+(const VarMat1& a, const VarMat2& b) {
+  check_matching_dims("operator+", "a", a, "b", b);
+  using op_ret_type = decltype(a.val() + b.val());
+  using ret_type = promote_var_matrix_t<op_ret_type, VarMat1, VarMat2>;
+  arena_t<ret_type> ret((a.val() + b.val()).eval());
+  arena_t<VarMat1> arena_a = a;
+  arena_t<VarMat2> arena_b = b;
+  reverse_pass_callback([ret, arena_a, arena_b]() mutable {
+    arena_a.adj() += ret.adj_op();
+    arena_b.adj() += ret.adj_op();
+  });
+  return ret_type(ret);
+}
+
+/**
+ * Addition operator for a matrix variable and arithmetic (C++).
+ *
+ * @tparam VarMat A matrix of vars or a var with an underlying matrix type.
+ * @tparam Arith A type with an arithmetic Scalar type.
+ * @param a First variable operand.
+ * @param b Second variable operand.
+ * @return Variable result of adding two variables.
+ */
+template <typename Arith, typename VarMat,
+          require_st_arithmetic<Arith>* = nullptr,
+          require_rev_matrix_t<VarMat>* = nullptr>
+inline auto operator+(const VarMat& a, const Arith& b) {
+  if (is_eigen<Arith>::value) {
+    check_matching_dims("operator+", "a", a, "b", b);
+  }
+  using op_ret_type
+      = decltype((a.val().array() + as_array_or_scalar(b)).matrix());
+  using ret_type = promote_var_matrix_t<op_ret_type, VarMat>;
+  if (is_equal(b, 0.0)) {
+    return ret_type(a);
+  } else {
+    arena_t<VarMat> arena_a = a;
+    arena_t<ret_type> ret(a.val().array() + as_array_or_scalar(b));
+    reverse_pass_callback(
+        [ret, arena_a]() mutable { arena_a.adj() += ret.adj_op(); });
+    return ret_type(ret);
   }
-  return {new internal::add_vd_vari(b.vi_, a)};  // by symmetry
+}
+
+/**
+ * Addition operator for an arithmetic type and matrix variable (C++).
+ *
+ * @tparam VarMat A matrix of vars or a var with an underlying matrix type.
+ * @tparam Arith A type with an arithmetic Scalar type.
+ * @param a First variable operand.
+ * @param b Second variable operand.
+ * @return Variable result of adding two variables.
+ */
+template <typename Arith, typename VarMat,
+          require_st_arithmetic<Arith>* = nullptr,
+          require_rev_matrix_t<VarMat>* = nullptr>
+inline auto operator+(const Arith& a, const VarMat& b) {
+  return b + a;
+}
+
+/**
+ * Addition operator for an arithmetic matrix and variable (C++).
+ *
+ * @tparam Var A `var_value` with an underlying arithmetic type.
+ * @tparam EigMat An Eigen Matrix type with an arithmetic Scalar type.
+ * @param a First variable operand.
+ * @param b Second variable operand.
+ * @return Variable result of adding two variables.
+ */
+template <typename Var, typename EigMat,
+          require_eigen_vt<std::is_arithmetic, EigMat>* = nullptr,
+          require_var_vt<std::is_arithmetic, Var>* = nullptr>
```

Comment:
Not related to this pull but I didn't realise we had this ```require_var_vt```, that will simplify some of the wacky templating I've been resorting to. These require generics have been such a good addition!